client/protocols/hinting.py

```diff
@@ -0,0 +1,185 @@
+from client.utils import auth
+from client.utils import format
+
+import json
+import logging
+import os
+import pickle
+import random
+import re
+import urllib.error
+import urllib.request
+
+log = logging.getLogger(__name__)
+
+#####################
+# Hinting Mechanism #
+#####################
+
+class HintingProtocol(protocol_models.Protocol):
+    """A protocol that provides rubber duck debugging and hints if applicable.
+    """
+
+    HINT_SERVER = "http://146.148.59.194:5000/"
+    HINT_ENDPOINT = 'api/hints'
+
+    def run(self, messages):
+        """Determine if a student is elgible to recieve a hint. Based on their
+        state, poses reflection questions.
+
+        After more attempts, ask if students would like hints. If so, query
+        the server.
+        """
+        if self.args.local:
+            return
+
+        if 'analytics' not in messages:
+            log.info('Analytics Protocol is required for hint generation')
+            return
+        if 'file_contents' not in messages:
+            log.info('File Contents needed to generate hints')
+            return
+
+        messages['hinting'] = {}
+        history = messages['analytics']['history']
+        questions = history['questions']
+        current_q = history['question']
+
+        for question in current_q:
+            if question not in questions:
+                continue
+            stats = questions[question]
+            messages['hinting'][question] = {'prompts': {}, 'reflection': {}}
+            hint_info = messages['hinting'][question]
+
+            # Only prompt
+            if (stats['solved'] and stats['attempts'] > 6):
+                hint_info['elgible'] = False
+                if self.args.question:
+                    # Only prompt for reflection with question specified.
+                    log.info('Giving reflection response on %s', question)
+                    reflection = random.choice(SOLVE_SUCCESS_MSG)
+                    if not confirm("Nice work! Could you answer a quick question"
+                                   " about how you approached this question?"):
+                        hint_info['reflection']['accept'] = False
+                    else:
+                        hint_info['reflection']['accept'] = True
+                    prompt_user(reflection, hint_info)
+            elif (stats['attempts'] < 5):
+                log.info("Question %s is not elgible: Attempts: %s, Solved: %s",
+                         question, stats['attempts'], stats['solved'])
+                hint_info['elgible'] = False
+            else:
+                hint_info['elgible'] = not stats['solved']
+
+            if not hint_info['elgible']:
+                continue
+            log.info('Prompting for hint on %s', question)
+
+            if confirm("Check for hints on {}?".format(question)):
+                hint_info['accept'] = True
+                print("Thinking... (could take upto 15 seconds)", end='')
```

Comment:
The idea is that the next print uses `\r` to clear the line. If we terminate the line we can't clear this line. 

In my experience it's been a little iffy at actually displaying this line if it doesn't end in a new line - so I'll change it back to end in a new line and avoid the line clearing business.